pkg/daemon/daemon.go

```diff
@@ -119,19 +127,158 @@ func (d *daemon) Run() {
+func (d *daemon) mapGUIDToPodNetwork(allocatedGUID, podNetworkID string, podUID types.UID) error {
+	if _, exist := d.guidPodNetworkMap[allocatedGUID]; exist {
+		if podNetworkID != d.guidPodNetworkMap[allocatedGUID] {
+			return fmt.Errorf("failed to allocate requested guid %s, already allocated for %s",
+				allocatedGUID, d.guidPodNetworkMap[allocatedGUID])
+		}
+	} else if err := d.guidPool.AllocateGUID(allocatedGUID); err != nil {
+		return fmt.Errorf("failed to allocate GUID for pod ID %s, wit error: %v", podUID, err)
+	} else {
+		d.guidPodNetworkMap[allocatedGUID] = podNetworkID
+	}
+
+	return nil
+}
+
+// Allocate GUID if network don't have it and update Pod's networks annotation
+func (d *daemon) ProceedNetworkGUID(pw *PodWrapper, networkID string) error {
+	var guidAddr guid.GUID
+	allocatedGUID, err := utils.GetPodNetworkGUID(pw.ibNetwork)
+	podNetworkID := string(pw.pod.UID) + networkID
+	if err == nil {
+		// User allocated guid manually or Pod's network was rescheduled
+		err = d.mapGUIDToPodNetwork(allocatedGUID, podNetworkID, pw.pod.UID)
+		if err != nil {
+			return err
+		}
+
+		guidAddr, err = guid.ParseGUID(allocatedGUID)
+		if err != nil {
+			return fmt.Errorf("failed to parse user allocated guid %s with error: %v", allocatedGUID, err)
+		}
+	} else {
+		guidAddr, err = d.guidPool.GenerateGUID()
+		if err != nil {
+			return fmt.Errorf("failed to generate GUID for pod ID %s, wit error: %v", pw.pod.UID, err)
+		}
+
+		allocatedGUID = guidAddr.String()
+		err = d.mapGUIDToPodNetwork(allocatedGUID, podNetworkID, pw.pod.UID)
+		if err != nil {
+			return err
+		}
+
+		if err = utils.SetPodNetworkGUID(pw.ibNetwork, allocatedGUID); err != nil {
+			return fmt.Errorf("failed to set pod network guid with error: %v ", err)
+		}
+
+		// Update Pod's network annotation here, so if network will be rescheduled we wouldn't allocate it again
+		netAnnotations, err := json.Marshal(pw.networks)
+		if err != nil {
+			return fmt.Errorf("failed to dump networks %+v of pod into json with error: %v", pw.networks, err)
+		}
+
+		pw.pod.Annotations[v1.NetworkAttachmentAnnot] = string(netAnnotations)
+	}
+
+	// used GUID as net.HardwareAddress to use it in sm plugin which receive []net.HardwareAddress as parameter
+	pw.addr = guidAddr.HardWareAddress()
+	return nil
+}
+
+// Update and set Pod's network annotation
+func (d *daemon) UpdatePodNetworkAnnotation(pw *PodWrapper, removedList *[]net.HardwareAddr) error {
+	(*pw.ibNetwork.CNIArgs)[utils.InfiniBandAnnotation] = utils.ConfiguredInfiniBandPod
+
+	netAnnotations, err := json.Marshal(pw.networks)
+	if err != nil {
+		return fmt.Errorf("failed to dump networks %+v of pod into json with error: %v", pw.networks, err)
+	}
+
+	pw.pod.Annotations[v1.NetworkAttachmentAnnot] = string(netAnnotations)
+
+	if err := d.kubeClient.SetAnnotationsOnPod(pw.pod, pw.pod.Annotations); err != nil {
+		if !strings.Contains(strings.ToLower(err.Error()), "not found") {
```

Comment:
can we use error code and not relay on string "not found" that may chang

Reply:
k8s error have a "one word" descriptor for an error which these methods compare agains